[Enhancement] - Multiple adaptions in the exam scores view

FuchsDominik opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, editor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all newly inserted strings into English and German.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
We want to have some little adaptions to the exam scores page, since some aspects are not ideally

### Description
<!-- Describe your changes in detail -->

1. Removed some whitespaces from the exam information below the exam title (with number of points, exercise groups etc.)
2. Added percentage value of how many students passed (if grading key defined) and how many percentage actually submitted
3. Adapted the exercise average scores, made the bars smaller (less height), added a possibility to click on the bar to navigate into the scores page of the corresponding exercise
4. Added a new line in the exam information matrix which shows the standard grade deviation (only if grade scale defined and grades only consist of numeric values!)
5. Explicitly add points or grades to the matrix to distinguish between those

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->

1. Log in to Artemis
2. Navigate into the scores page of an exam (preferably with second correction done and grade key defined, in order to check everything)
3. Check the improvements mentioned in the Description section and if values are also correct (no unexpected values at least, if you want to calculate the standard deviation of the grades use [this](https://www.investopedia.com/terms/s/standarddeviation.asp))
4. Add a string to the grade scale names (instead of 5.0 for example) and make sure the standard grade deviation is not shown

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->
Table:
<img width="687" alt="Screenshot 2021-08-02 at 11 29 34" src="https://user-images.githubusercontent.com/44401560/127844186-8b9d47df-bda8-4c12-826a-5b5988437fc1.png">

Made graph bit thinner:
![Screenshot 2021-08-02 at 15 51 21](https://user-images.githubusercontent.com/44401560/127872638-01dbd46f-a4db-45c2-a82d-6532c2d4ef47.png)
Now:
<!-- ![Screenshot 2021-08-02 at 15 51 12](https://user-images.githubusercontent.com/44401560/127872650-0ce87ca7-f6e8-4e07-8947-b7d4d720886b.png) -->
<img width="814" alt="Screenshot 2021-08-28 at 13 35 53" src="https://user-images.githubusercontent.com/44401560/131216630-3afeb71c-002a-4225-adcf-27ad264dd5ad.png">


### Review Progress
<!-- Each Pull Request should be reviewed by at least two other developers. The code as well as the functionality (= manual test) needs to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->

- Code Review
  - [x] Review 1
  - [x] Review 2
- Manual Tests
  - [x] Test 1
  - [x] Test 2

Malyuk-A reviewed (approved):
Tested on TS1. Works as described. Code seems reasonable.

lschlesinger commented on src/main/webapp/app/exam/exam-scores/exam-scores.component.ts:
```diff
@@ -291,7 +295,8 @@ export class ExamScoresComponent implements OnInit, OnDestroy {
             exerciseGroup.containedExercises.forEach((exerciseInfo) => {
-                const exResult = new AggregatedExerciseResult(exerciseInfo.exerciseId, exerciseInfo.title, exerciseInfo.maxPoints, exerciseInfo.numberOfParticipants);
+                const type = ExamScoresComponent.declareExerciseType(exerciseInfo);
+                const exResult = new AggregatedExerciseResult(exerciseInfo.exerciseId, exerciseInfo.title, exerciseInfo.maxPoints, exerciseInfo.numberOfParticipants, type!);
```
I would not call it exResult, it could be exam or exercise

FuchsDominik replied:
It was already called exResult before, I only added a parameter to the constructor. But let's call it exerciseResult then.

lschlesinger commented on src/main/webapp/i18n/de/exam.json:
```diff
@@ -93,7 +93,7 @@
             "titleExerciseGroupColumn": "Aufgabengruppe",
             "titleExerciseColumn": "Aufgabe",
-            "averagePointsColumn": "Durchschnitt",
+            "averagePointsColumn": "Durchschnittspunktzahl",
```
Maybe "Durchschnittliche Punktzahl" (it's also Durchschnittliche Note / Bonuspunkte)

FuchsDominik replied:
Yes, sounds more consistent 👍

lschlesinger commented on src/main/webapp/app/exam/exam-scores/exam-scores-average-scores-graph.component.ts:
```diff
@@ -137,6 +163,42 @@ export class ExamScoresAverageScoresGraphComponent implements OnInit {
                     },
                 },
             },
+            // we show the a pointer to indicate to the user that a data point is clickable (navigation to exercise)
```
typo, the or a

FuchsDominik replied:
👍

lschlesinger commented on src/main/webapp/app/exam/exam-scores/exam-scores.component.ts:
```diff
@@ -713,4 +777,19 @@ export class ExamScoresComponent implements OnInit, OnDestroy {
         Sentry.captureException(new Error(errorMessage));
     }
+
+    private static declareExerciseType(exerciseInfo: ExerciseInfo) {
+        switch (exerciseInfo.exerciseType) {
+            case 'TextExercise':
+                return ExerciseType.TEXT;
+            case 'ModelingExercise':
+                return ExerciseType.MODELING;
+            case 'ProgrammingExercise':
+                return ExerciseType.PROGRAMMING;
+            case 'FileUploadExercise':
+                return ExerciseType.FILE_UPLOAD;
+            case 'QuizExercise':
+                return ExerciseType.QUIZ;
+        }
+    }
```
can't be initially set the type to an enum type in `exam-score-dtos.model.ts`? I am not sure about this, just a thought.

FuchsDominik replied:
We cannot get around mapping the actual Java Class name to the shorter name (quiz instead of QuizExercise) for the URL. Do you think mapping it server-side has an advantage?

lschlesinger replied:
no just leave it as it is... thanks for clarification, I was a bit confused, I think the way you implemented it is reasonable.

FuchsDominik replied:
👍

lschlesinger commented on src/main/webapp/app/exam/exam-scores/exam-score-dtos.model.ts:
```diff
@@ -25,6 +27,7 @@ export class ExerciseInfo {
     public title: string;
     public maxPoints: number;
     public numberOfParticipants: number;
+    public exerciseType: string;
```
Don't we have an enum type that represents the server-side enums?

FuchsDominik replied:
We do not have a server-side exercise type enum. I saw us using the actual Class name of the exercise which is then TextExercise for example, which is was I am doing in this case

lschlesinger reviewed (commented):
I have some comments regarding the code. Is there an exam prepared somewhere on a test server, for which I can review the changes in the statistics? Otherwise, I would have to create one and participate :D

FuchsDominik commented:
@lschlesinger Thanks for the feedback :) Simon has an exam in his course where I tested my changes, with test_user_13 I think

Malyuk-A reviewed (approved):
Re-approve (since my last review only a few local renamings took place)

Malyuk-A commented:
Concerning the failing TypeScript Client Test:
As I can see you have similar problems too ... (every test passes but still build failed)
Lorena mentioned that the issue might be with a test/line coverage that is not fulfilled. 
(Even if you only added few lines of code)
![image](https://user-images.githubusercontent.com/65814168/128760544-5c04df7e-d080-4aff-a69d-5ded4428bc2f.png)

Olimbia23 commented on src/main/webapp/app/exam/exam-scores/exam-scores-average-scores-graph.component.ts:
```diff
@@ -137,6 +163,42 @@ export class ExamScoresAverageScoresGraphComponent implements OnInit {
+    /**
+     * We navigate to the exercise scores page when the user clicks on a data point
+     */
+    navigateToExercise(exerciseId: number, exerciseType: ExerciseType) {
+        this.router.navigate([
+            'course-management',
+            this.courseId,
+            'exams',
+            this.examId,
+            'exercise-groups',
+            this.averageScores.exerciseGroupId,
+            `${exerciseType}-exercises`,
+            exerciseId,
+            'scores',
+        ]);
+    }
```
[nit] What do you think of moving this to `navigation-utils`? There are some functions similar to this: https://github.com/ls1intum/Artemis/blob/master/src/main/webapp/app/utils/navigation.utils.ts#L71-L91

FuchsDominik replied:
Yeah that's actually a good idea, I created a generic method in the navigation utils to navigate into subpages of exam exercises

Olimbia23 commented on src/main/webapp/app/exam/exam-scores/exam-scores.component.ts:
```diff
@@ -703,4 +767,19 @@ export class ExamScoresComponent implements OnInit, OnDestroy {
         Sentry.captureException(new Error(errorMessage));
     }
+
+    private static declareExerciseType(exerciseInfo: ExerciseInfo) {
+        switch (exerciseInfo.exerciseType) {
+            case 'TextExercise':
+                return ExerciseType.TEXT;
+            case 'ModelingExercise':
+                return ExerciseType.MODELING;
+            case 'ProgrammingExercise':
+                return ExerciseType.PROGRAMMING;
+            case 'FileUploadExercise':
+                return ExerciseType.FILE_UPLOAD;
+            case 'QuizExercise':
+                return ExerciseType.QUIZ;
+        }
```
[nit] I think we should create a build/make/construct function for `ExerciseType` enum.
Something like: `ExerciseType.fromExerciseInfo(exerciseInfo: ExerciseInfo) { ... }`

It seems like in typescript one would do it like this: https://stackoverflow.com/questions/28150739/add-functions-to-an-enum
What do you think?

FuchsDominik replied:
How about simply creating a `createExerciseTypeByExerciseInfo` method in the `exercise.model.ts`? This way we still have the enum

Olimbia23 reviewed (dismissed):
Reviewed the code, left minor suggestions.
Overall, good work. I approve it @FuchsDominik 👍

FuchsDominik commented:
@Olimbia23 Thanks for the feedback. I adapted the code like you suggested :)

Malyuk-A reviewed (approved):
Re-approve

Olimbia23 reviewed (dismissed):
Thank you for applying the suggestions @FuchsDominik. 
Approve 👍

lschlesinger reviewed (commented):
I have some general questions and remarks, I am not sure if you share my opinion and whant to address it, that's why I first put the review as a comment :) Let me know, what you think. Here are my notes when I inspected the page from top to bottom:
- I think the german translation for characteristics right below the exam name are a bit strange "40 Maximale Punktzahl", you could change for both languages to: Maximale Punktzahl/Maximum Points: 40 etc.
- why are there no values in for passed/submitted/total exams in first correction (would't it be interesting to see, if with the second correction more/less students failed?)
- why did you not include the column "not submitted", then the total column would be the sum of the other three columns. This would be intuitive if total can be calculated out of all the other information in the table
- The german translation for the last 4 rows could be "Punkte Median", "Note Median", "Punkte Standardabweichung", "Note Standardabweichung", this is more consistent than using the dash and writing it as one word
- The legend of the histogram behaves a bit strange when it comes to changing the language. I managed to get the whole page in english but "Anzahl Studierende" and "Ergebnis %" in german for the axes, maybe that could be fixed, I don't know.
- I actually don't think the legend (grey rectangle with # of students) is needed, as the axes already says "number of participants", actually it's the same but you now have 2 slightly different explanations
- Am I missing the explanation for the exam groups and exam statistics? what do the two differently coloured bars express? IMO there is missing the information
- The german translation for exam groups and exam statistics is "Aufgabengruppen- & Aufgabenstatistiken", what is the dash needed for? Should it be "Aufgabengruppen & -statistik"?

FuchsDominik commented:
@lschlesinger Thanks for the feedback. It seems most of the things you mentioned relate to the exam scores page itself, not to the changes I did on this page (e.g. this PR only adds the last line to the exam information table, not the whole table ;)). Nevertheless there are definitely some valid points you mentioned

1. Works in english, sounds a bit strange in german, I agree
2. You are right, the number of passed students in the first correction round would be interesting (submitted and total exams are left blank as it is the same as for the second correction, but I could maybe just add those number so nobody gets confused)
3. I can just guess that this was thought be the most uninteresting aspect without added value and therefore was left unmentioned
4. Agree, will be changed
5. I experienced this in other places as well. Problem is that the chart itself is more like a blackbox and is handled by the initialization of the chart. In combination with `changeDetection: ChangeDetectionStrategy.OnPush` which we use on that page, the language change is not detected and therefore not changed. I could look into this, but probably not in this PR.
6. That's true, not necessarily needed
7. I tried to make the difference as clear as possible. The title of each average score graph, together with the same title on the left side of the bar and the blue bar should indicate that the first (blue) bar shows the average score of the exercise group, while the grey bars below show the average scores of the specific exercises, with the title of the exercise on the left hand side. Any idea how I can make this even more obvious?
8. It's supposed to be called Aufgabengruppenstatistiken & Aufgabenstatistiken, that's why I called it Aufgabengruppen- & Aufgabenstatistiken.

lschlesinger commented:
Thanks for addressing the comments!
> I tried to make the difference as clear as possible. The title of each average score graph, together with the same title on the left side of the bar and the blue bar should indicate that the first (blue) bar shows the average score of the exercise group, while the grey bars below show the average scores of the specific exercises, with the title of the exercise on the left hand side. Any idea how I can make this even more obvious?

Maybe:
e.g. right aligning the name of the exercise group / exercise, I think the short name was a bit lost when it was left aligned. Also what is the number before the exercise title? It was not displayed anywhere else, could be just drop it. It would indicate, that these are average points in comparison by changing the title and the the bar descriptions such as shown in the picture. I did not see the tooltip before, and without the tooltip, one would not now that it's about average scores.
<img width="1433" alt="Bildschirmfoto 2021-08-28 um 10 31 47" src="https://user-images.githubusercontent.com/41366522/131211842-d405a87c-c7b7-46c7-9d19-735094cbd776.png">

FuchsDominik commented:
> Thanks for addressing the comments!
> 
> > I tried to make the difference as clear as possible. The title of each average score graph, together with the same title on the left side of the bar and the blue bar should indicate that the first (blue) bar shows the average score of the exercise group, while the grey bars below show the average scores of the specific exercises, with the title of the exercise on the left hand side. Any idea how I can make this even more obvious?
> 
> Maybe:
> e.g. right aligning the name of the exercise group / exercise, I think the short name was a bit lost when it was left aligned. Also what is the number before the exercise title? It was not displayed anywhere else, could be just drop it. It would indicate, that these are average points in comparison by changing the title and the the bar descriptions such as shown in the picture. I did not see the tooltip before, and without the tooltip, one would not now that it's about average scores.
> <img alt="Bildschirmfoto 2021-08-28 um 10 31 47" width="1433" src="https://user-images.githubusercontent.com/41366522/131211842-d405a87c-c7b7-46c7-9d19-735094cbd776.png">

The number before the exercise title is the id of the exercise, Stephan requested it since there could be the possibility to have identical exercise names. You are very right that there is actually no indicator that we look at average scores here. I would propose to call the title of the graph `Average scores of exercise group TITLE in comparison`, therefore we indicate what we actually show while also explaining what the exercise group title is called, which will help noticing that the exercise group average is also shown. Regarding the right alignment of the bar title I am not that persuaded, because I think it's important to have the IDs aligned below each other which makes it easier to compare them

Olimbia23 reviewed (approved):
Reapprove 👍

lschlesinger reviewed (approved):
Looks good now 👍

Malyuk-A reviewed (approved):
re-approve. Tested on TS3
